src/box/box.cc

```diff
@@ -681,6 +691,194 @@ box_check_uri(const char *source, const char *option_name)
+		value = parse_uri_transport_option_value(val);
+		break;
+	/* key was checked before */
+	default:
+		unreachable();
+	}
+	if (value == URI_OPTION_VALUE_UNKNOWN) {
+		diag_set(ClientError, ER_CFG, "listen",
+			 tt_sprintf("invalid value `%s` for "
+				    "URI `%s` option", val, opt));
+		return -1;
+	}
+	return 0;
+}
+
+/**
+ * This function expects a string containing one or more URI options
+ * separated by "&". Each option must have a value separated by '='.
+ * Extra delimiters are ignored. "option1=value1&option2=valu2".
+ * Function continues the partitioning and checks started in function
+ * `box_check_uri_with_options`.
+ */
+static int
+box_check_uri_options(char *opts)
+{
+	static const char *error = "unable to parse URI options, extra `=`";
+	char *saveptr1, *saveptr2, *saveptr3;
+	for (char *optval = opts; opts != NULL; optval = NULL) {
+		char *optval_token = strtok_r(optval, "&", &saveptr1);
+		if (optval_token == NULL)
+			break;
+		char *opt = strtok_r(optval_token, "=", &saveptr2);
+		char *vals = strtok_r(NULL, "=", &saveptr2);
+		if (opt == NULL) {
+			diag_set(ClientError, ER_CFG, "listen",
+				 "invalid value `null` for URI option");
+			return -1;
+		}
+		if (vals == NULL) {
+			diag_set(ClientError, ER_CFG, "listen",
+				 tt_sprintf("invalid value `null` for URI "
+					    "`%s` option", opt));
+			return -1;
+		}
+		if (box_check_end_of_token("=", &saveptr2, error) != 0)
+			return -1;
+		for (char *val = vals; vals != NULL; val = NULL) {
+			char *val_token = strtok_r(val, ";", &saveptr3);
+			if (val_token == NULL)
+				break;
+			if (box_check_uri_option(opt, val_token) != 0)
+				return -1;
+		}
+	}
+	return 0;
+}
+
+/**
+ * This function expects a string containing one or more URI separated
+ * by commas. Each URI can contain options separated by '?'. Each option
+ * is separated from the next one by '&'. Extra delimiters are ignored.
+ * "uri1?option1=value1&option2=value2, uri2?option1=value1&option2=value2"
+ */
+static int
+box_check_uri_with_options(const char *source)
+{
+	int rc = 0;
+	char *uri_with_opts, *uri_with_opts_token;
+	char *saveptr1, *saveptr2;
+	static const char *uri_opt_error = "unable to parse URI options, extra `?`";
+
+	char *source_copy = strdup(source);
+	if (source_copy == NULL) {
+		diag_set(OutOfMemory, strlen(source) + 1,
+			 "strdup", "source");
+		return -1;
+	}
+
+	rc = box_check_uri_delimiters(source_copy);
+	if (rc != 0)
+		goto cleanup;
+
+	for (uri_with_opts = source_copy; ; uri_with_opts = NULL) {
+		uri_with_opts_token =
+			strtok_r(uri_with_opts, ", ", &saveptr1);
+		if (uri_with_opts_token == NULL)
+			break;
+		rc = box_check_uri_delimiters(uri_with_opts_token);
+		if (rc != 0)
+			goto cleanup;
+		char *uri = strtok_r(uri_with_opts_token, "?", &saveptr2);
+		char *opts = strtok_r(NULL, "?", &saveptr2);
+		rc = box_check_end_of_token("?", &saveptr2, uri_opt_error);
+		if (rc != 0)
+			goto cleanup;
+		if ((rc = box_check_uri(uri, "listen")) != 0)
```

Comment:
You could try to use `uri_parse` a bit earlier. It seems it could do for you the job of separating the query parameters by `?`. Look at `struct uri.query`. Isn't it what you need?